File: Convex-Optimization-Fall-2017/homework-3-solution.py
# -*- coding: utf-8 -*-
"""
Spyder Editor

This is a temporary script file.
"""

import numpy as np
import cvxpy as cvx


m = 30
n = 100
A_real = np.random.randn(m, n)
A_imag = np.random.randn(m, n)
b_real = np.random.randn(m, 1)
b_imag = np.random.randn(m, 1)

A_temp1 = np.hstack( ( A_real, -A_imag ) )
A_temp2 = np.hstack( ( A_imag, A_real ) )

A_tilde = np.vstack( ( A_temp1, A_temp2 ) )
b_tilde = np.vstack( ( b_real, b_imag ) )


z = cvx.Variable(2*n,1)
t = cvx.Variable(1,1)


constraints = [
                A_tilde * z == b_tilde
              ]

for i in range(0, n-1, 1):
    ci = np.zeros( (2, 2*n) )
    ci[0][i] = 1
    ci[1][n+i] = 1
    soc_constraint = (cvx.norm( ci*z ) <= t)
    constraints += [ soc_constraint ]

obj = Minimize( t )


prob = Problem(obj, constraints)

result = prob.solve()

print(result)

File: Convex-Optimization-Fall-2017/homework-3-solution.m
% EE 617, Fall 2017, Homework 3

% generate problem data
m = 30;
n = 100;
A = randn( m , n ) + 1i * randn( m , n );
b = randn( m , 1 ) + 1i * randn( m , 1 );

% solve the problem directly
cvx_begin
    variable x(n) complex;
    minimize norm( x , inf )
    subject to 
        A * x == b
cvx_end

optimal_value_complex = cvx_optval;

% solve the SOCP with real variables
A_tilde = [ real(A) -imag(A) ; imag(A) real(A) ];
b_tilde = [ real(b) ; imag(b) ];

C = zeros( 2 , 2*n , n );

for i=1:1:n
    C(:,:,i) = zeros( 2 , 2*n );
    C( 1 , i , i ) = 1;
    C( 2 , n+i , i ) = 1;
end

cvx_begin
    variables t z(2*n);
    minimize t
    subject to 
        for i=1:1:n
            norm( C(:,:,i) * z , 2 ) <= t;
        end
        A_tilde * z == b_tilde;
cvx_end

optimal_value_real = cvx_optval;